### PR TITLE
Light Cell Geometry

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

This PR changes the data structure for storing cell covers in CellGeometry. Previous one was a CubicalComplex, which is based on a std::unordered_map Cell -> Data (here Data is not used). We use instead a new class UnorderedSetByBlock to store set of Points (Cells are represented by their Khalimsky points). The UnorderedSetByBlock is an std::unordered_map Point -> Word, where points are gathered by blocks of 32 consecutive points along x axis. If any point in a block is present, the corresponding element is present and bits to 1 in the word correspond to points present in the set. On average, memory occupancy is divided by four, the structure is four times faster for traversal and approximately same speed for queries/insertion/erase. 
CellGeometry and DigitalConvexity use now this data structure. A drawback is that this version is not fully backward compatible with the first version since the few operations involving CubicalComplex are removed.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: tests/kernel/testUnorderedSetByBlock.cpp [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+using namespace std;
+using namespace DGtal;
+
+template < typename Point >
+Point randomPoint( int S )
+{
+  Point p;
+  for ( DGtal::Dimension i = 0; i < Point::dimension; i++ )
+    p[ i ] = random() % S;
```
Comment (rolanddenis) [inline tests/kernel/testUnorderedSetByBlock.cpp:48]: Where did this `random` function come from?!!

Oh it seems to be a Linux-specific function, not in the C++ specification and thus not available with Windows. Could you replace it by `rand` or use the standard library?

BTW, could it be possible to also test negative coordinates?

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 2]
```diff
@@ -0,0 +1,1218 @@
...
+  public:
+    // ---------------------- iterators --------------------------------
+    
+    /// Read iterator on set elements. Model of ForwardIterator.
+    struct const_iterator
+      : public boost::iterator_facade< const_iterator, Key const,
+				       boost::forward_traversal_tag,
+				       Key const >
+    {
+      friend class UnorderedSetByBlock< Key, TSplitter, Hash, KeyEqual >;
```
Comment (rolanddenis) [inline src/DGtal/kernel/UnorderedSetByBlock.h:331]:
```suggestion
      friend struct UnorderedSetByBlock< Key, TSplitter, Hash, KeyEqual >;
```
or define `UnorderedSetByBlock` as a `class`.

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 3]
```diff
@@ -0,0 +1,1218 @@
...
+  /// 4 x times faster !
+  ///
+  /// @note The generic class is not implemented since it is based on bit operations.
+  ///
+  /// @see UnorderedSetByBlock
+  template < typename TElement >
+  struct Splitter {
+    typedef Splitter< TElement > Self;
+    typedef TElement             Element;
+    typedef uint32_t             Word;
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:61]: I should remove type Word from Splitter, since Words are used only in UnorderedSetByBlock

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 4]
```diff
@@ -0,0 +1,1218 @@
...
+    ///
+    /// @return a pair grouping its block coordinates and its bit
+    /// within this block.
+    static
+    std::pair< Element, DGtal::Dimension >
+    split( const Element& e )
+    {
+      Element se = e;
+      se[ 0 ]   &= 0xffffffffffffffe0ll;
+      return std::make_pair( se, (DGtal::Dimension) e[ 0 ] & 0x000000000000001fll );
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:217]: I was more confident with bitwise operations. I thought /32 and %32 would work fine: they did but were 4 times slower. Do you think your representation is clearer ?
Otherwise, yes 64 bit word could be considered and is would be cheaper on average. But I was lazy I guess. In a forthcoming PR perhaps, because it would require some benchmarking probably ?

File: src/DGtal/kernel/UnorderedSetByBlock.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (rolanddenis) [reply src/DGtal/kernel/UnorderedSetByBlock.h:61]: Yes but it makes sense that Splitter "choose" the right type to manipulate bitset depending on the block size! I was just wondering if the coordinates inside the block should also use the same type even if the constraints are not the same (but it has more meaning for me than a `DGtal::Dimension`)

File: src/DGtal/kernel/UnorderedSetByBlock.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (rolanddenis) [reply src/DGtal/kernel/UnorderedSetByBlock.h:217]: It is not about the bitwise operations but specifically mixing signed and unsigned integers, like the `0xffffffffffffffe0` mask stored in a signed integer. I completely agree that using integer division and modulo will be slower and, BTW, will not handle negative coordinates appropriately (e.g. -40 will be mapped to the -32 block with -8 coordinate) without additional test.

I was more thinking to the version that uses the decimal representation and a subtraction (instead of a mask) to get the inside coordinates, e.g.:
```C++
static
std::pair< Element, DGtal::Dimension >
split( Element e )
{
    auto block_coords = static_cast<DGtal::Dimension>(e[ 0 ] & static_cast<typename Element::Coordinate>(31));
    e[ 0 ] -= block_coords;    
    return { e, block_coords };
}
```

I reused `Element` that is passed by copy instead of reference. Maybe to much `static_cast` but it is to avoid unwanted implicit conversions (and associated warnings). The advantages are that the block size is thus more easily customizable (for later, e.g. using `std::bitset` that should be equivalent to using `uintN_t` for N <= 64) and there is not need to have two specialized versions (still need to detect if a fundamental type is used or note :thinking: )

File: src/DGtal/geometry/volumes/CellGeometry.ih [hunk 5]
```diff
@@ -269,70 +257,52 @@ maxCellDim() const
...
-    }
+  UnorderedSetByBlock< Point > k_dim_points;
+  for ( auto&& c : myKPoints )
+    if ( dim( c ) == k )
+      k_dim_points.emplace( c );
+  return other.myKPoints.includes( k_dim_points );
+  // for ( auto&& c : myKPoints )
+  //   if ( dim( c ) == k && other.myKPoints.count( c ) == 0 )
+  //     return false;
+  // return true;
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/CellGeometry.ih:286]: done

File: src/DGtal/geometry/volumes/DigitalConvexity.ih [hunk 6]
```diff
@@ -306,15 +306,25 @@ DGtal::DigitalConvexity<TKSpace>::
...
-      if ( ( intersected_cells.nbCells( k ) != touched_cells.nbCells( k ) )
-	   || ( ! intersected_cells.subset( touched_cells, k ) ) )
+      if ( ( intersected_cells.nbCells() != touched_cells.nbCells() )
+	   || ( ! intersected_cells.subset( touched_cells ) ) )
 	return false;
     }
   return true;
+  // auto S = insidePoints( P );
+  // auto touched_cells     = makeCellCover( S.begin(), S.end(), 1, KSpace::dimension - 1 );
+  // for ( Dimension k = 1; k < KSpace::dimension; ++ k )
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/DigitalConvexity.ih:320]: removed

File: src/DGtal/geometry/volumes/DigitalConvexity.ih [hunk 7]
```diff
@@ -332,14 +342,16 @@ bool
...
-  for ( Dimension k = C.minCellDim(); k <= C.maxCellDim(); ++k )
-    {
-      auto intersected_cells = makeCellCover( P, k, k );
-      if ( C_prepared ) intersected_cells.prepareSubsetOperations();
-      if ( ! intersected_cells.subset( C ) ) return false;
-    }
-  return true;
+  auto intersected_cells = makeCellCover( P, C.minCellDim(), C.maxCellDim() );
+  return intersected_cells.subset( C );
+  // bool C_prepared = C.isPreparedForSubsetOperations();
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/DigitalConvexity.ih:347]: done

File: src/DGtal/geometry/volumes/DigitalConvexity.ih [hunk 8]
```diff
@@ -363,15 +375,25 @@ DGtal::DigitalConvexity<TKSpace>::
...
       auto intersected_cells = makeCellCover( P, k, k );
-      if ( ( intersected_cells.nbCells( k ) != touched_cells.nbCells( k ) )
-	   || ( ! intersected_cells.subset( touched_cells, k ) ) )
+      if ( ( intersected_cells.nbCells() != touched_cells.nbCells() )
+	   || ( ! intersected_cells.subset( touched_cells ) ) )
 	return false;
     }
   return true;
+  // auto S = insidePoints( P );
+  // auto touched_cells     = makeCellCover( S.begin(), S.end(), 1, KSpace::dimension - 1 );
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/DigitalConvexity.ih:388]: done

File: src/DGtal/geometry/volumes/DigitalConvexity.ih [hunk 9]
```diff
@@ -390,14 +412,16 @@ bool
...
-      auto intersected_cells = makeCellCover( P, k, k );
-      if ( C_prepared ) intersected_cells.prepareSubsetOperations();
-      if ( ! intersected_cells.subset( C ) ) return false;
-    }
-  return true;
+  auto intersected_cells = makeCellCover( P, C.minCellDim(), C.maxCellDim() );
+  return intersected_cells.subset( C );
+  // bool C_prepared = C.isPreparedForSubsetOperations();
+  // for ( Dimension k = C.minCellDim(); k <= C.maxCellDim(); ++k )
+  //   {
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/DigitalConvexity.ih:419]: done

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 10]
```diff
@@ -0,0 +1,1218 @@
...
+    }
+  }; // struct Splitter< PointVector< dim, DGtal::int64_t, TContainer > >
+
+  /// This data structure represents a set of elements that must be
+  /// integral arrays (i.e. digital points). It is similar to an
+  /// unordered_set, but much more compact in general. The container
+  /// used is a unordered_map from point to a 32-bit word. The idea is
+  /// to group consecutive elements/points along x axis by blocks of
+  /// 32 bits.  If any point in a block is present, the corresponding
+  /// element is present in the unordered_map and bits to 1 in the
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:257]: "bits sets to 1"

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 11]
```diff
@@ -0,0 +1,1218 @@
...
+	  my_size     = other.my_size;
+	}
+      return *this;
+    }
+    /// Default move assignment
+    /// @param other the object to clone
+    /// @return a reference to this
+    Self& operator=( Self&& other )
+    {
+      if ( this != &other )
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:634]: fixed

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 12]
```diff
@@ -0,0 +1,1218 @@
...
+  /// @tparam KeyEqual the type that provides an equality comparator for Key.
+  /// @tparam UnorderedMapAllocator the type that provides an allocator for the underlying unordered_map container.
+  template < typename Key,
+	     typename TSplitter,
+	     class Hash,
+	     class KeyEqual,
+	     class UnorderedMapAllocator >
+  void swap
+  ( UnorderedSetByBlock< Key, TSplitter, Hash, KeyEqual, UnorderedMapAllocator > s1,
+    UnorderedSetByBlock< Key, TSplitter, Hash, KeyEqual, UnorderedMapAllocator > s2 )
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:1210]: Oh yeah ! What a mistake !

File: tests/kernel/testUnorderedSetByBlock.cpp (hunk 1 above)
Comment (JacquesOlivierLachaud) [reply tests/kernel/testUnorderedSetByBlock.cpp:48]: I have replaced random by rand and check also negative coords

File: src/DGtal/kernel/UnorderedSetByBlock.h (hunk 2 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:331]: done

File: src/DGtal/kernel/UnorderedSetByBlock.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:217]: Ok. I see your point. I modify the code and also allows the user to choose word type

File: src/DGtal/kernel/UnorderedSetByBlock.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:61]: ok

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 13]
```diff
@@ -209,12 +80,13 @@ namespace DGtal
...
-    std::pair< Element, DGtal::Dimension >
-    split( const Element& e )
+    std::pair< Element, Coordinate >
+    split( Element e )
     {
-      Element se = e;
-      se[ 0 ]   &= 0xffffffffffffffe0ll;
-      return std::make_pair( se, (DGtal::Dimension) e[ 0 ] & 0x000000000000001fll );
+      auto block_coords = 
+	( e[ 0 ] & static_cast<Coordinate>( sizeof( Word ) * 8 - 1 ) );
```
Comment (rolanddenis) [inline src/DGtal/kernel/UnorderedSetByBlock.h:87]:
```suggestion
      auto block_coords = 
        ( e[ 0 ] & static_cast<Coordinate>( sizeof( Word ) * CHAR_BITS - 1 ) );
```
:ok_hand:

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 14]
```diff
@@ -30,177 +30,48 @@
 #include <boost/iterator/iterator_facade.hpp>
```
Comment (rolanddenis) [inline src/DGtal/kernel/UnorderedSetByBlock.h:30]:
```suggestion
#include <boost/iterator/iterator_facade.hpp>
#include <climits>
```
For `CHAR_BITS` used below.

File: src/DGtal/kernel/UnorderedSetByBlock.h [hunk 15]
```diff
@@ -30,177 +30,48 @@
...
-    typedef Splitter< Element >  Self;
-    typedef uint32_t             Word;
+  template < typename TElement, typename TWord = uint32_t >
+  struct Splitter {
+    typedef Splitter< TElement,TWord >   Self;
+    typedef TElement                     Element;
+    typedef TWord                        Word;
+    typedef typename Element::Coordinate Coordinate;
 
+    BOOST_CONCEPT_ASSERT(( concepts::CIntegralNumber< Word > ));
```
Comment (rolanddenis) [inline src/DGtal/kernel/UnorderedSetByBlock.h:72]:
```suggestion
    BOOST_CONCEPT_ASSERT(( concepts::CBoundedNumber< Word > ));
```
To remove `BigInteger` from the accepted word types.

Comment (JacquesOlivierLachaud) [reply src/DGtal/kernel/UnorderedSetByBlock.h:73]: Tou are right !

**Comment by JacquesOlivierLachaud** [comment]
Finally it works. I do not know where was the problem (since it was working fine on my laptop), so I static_cast everywhere possible. There was an implicit conversion that went bad on some system certainly. So I think the PR is ready now.

**Comment by dcoeurjo** [comment]
thx merging
